File: src/main/java/study/codingtest/leetcode/LC_239.kt
fun maxSlidingWindow(nums: IntArray, k: Int): IntArray {
  val result = mutableListOf<Int>()
  val dq = ArrayDeque<Int>()

  for (i in nums.indices) {
    if (dq.isNotEmpty() && dq.first() < i - k + 1) {
      dq.removeFirst()
    }
    while (dq.isNotEmpty() && nums[dq.last()] < nums[i]) {
      dq.removeLast()
    }
    dq.addLast(i)
    if (i >= k - 1) {
      result.add(nums[dq.first()])
    }
  }
  return result.toIntArray()
}

File: src/main/java/study/codingtest/leetcode/LC_239.ts
const maxSlidingWindow = (nums: number[], k: number): number[] => {
  const result: number[] = [];
  const dq: number[] = [];

  for (let i = 0; i < nums.length; i++) {
    if (dq.length && dq[0] <= i - k) {
      dq.shift();
    }
    while (dq.length && nums[dq[dq.length - 1]] <= nums[i]) {
      dq.pop();
    }
    dq.push(i);
    if (i >= k - 1) {
      result.push(nums[dq[0]]);
    }
  }
  return result;
};
